python_scripts/linear_models_sol_03.py

```diff
@@ -59,64 +62,239 @@
+import warnings
+
+
+warnings.filterwarnings("ignore", module="seaborn")
 
 for C in Cs:
     logistic_regression.set_params(logisticregression__C=C)
     logistic_regression.fit(data_train, target_train)
     accuracy = logistic_regression.score(data_test, target_test)
 
-    DecisionBoundaryDisplay.from_estimator(
+    disp = DecisionBoundaryDisplay.from_estimator(
         logistic_regression,
-        data_test,
-        response_method="predict",
+        data_train,
+        response_method="predict_proba",
+        plot_method="pcolormesh",
         cmap="RdBu_r",
-        alpha=0.5,
+        alpha=0.8,
+        # Setting vmin and vmax to the extreme values of the probability in
+        # order to make sure that 0.5 is mapped to white (the middle) of the
+        # blue-red colormap.
+        vmin=0.0,
+        vmax=1.0,
+    )
+    DecisionBoundaryDisplay.from_estimator(
+        logistic_regression,
+        data_train,
+        response_method="predict_proba",
+        plot_method="contour",
+        linestyles="--",
+        linewidths=1,
+        alpha=0.8,
+        levels=[0.5],  # 0.5 probability contour line
+        ax=disp.ax_,
     )
     sns.scatterplot(
-        data=penguins_test,
+        data=penguins_train,
         x=culmen_columns[0],
         y=culmen_columns[1],
         hue=target_column,
-        palette=["tab:red", "tab:blue"],
+        palette=["tab:blue", "tab:red"],
+        ax=disp.ax_,
     )
     plt.legend(bbox_to_anchor=(1.05, 0.8), loc="upper left")
     plt.title(f"C: {C} \n Accuracy on the test set: {accuracy:.2f}")
 
+# %% [markdown] tags=["solution"]
+#
+# On this series of plots we can observe several important points:
+#
+# - the darker the color, the more confident the classifier is in its
+#   predictions. Indeed, the darker the color, the closer the predicted
+#   probability is to 0 or 1;
+# - for lower values of `C` (stronger regularization), the classifier is less
+#   confident in its predictions (the near-white area is larger);
+# - for higher values of `C` (weaker regularization), the classifier is more
+#   confident: the areas with dark blue (very confident in predicting Adelie)
+#   and dark red (very confident in predicting Chinstrap) nearly cover the
+#   entire feature space;
+# - the direction of the straight line separating the two classes is impacted
+#   by the choice of `C`: the smaller `C` (the higher the regularization), the
+#   more the decision bounday is influenced almost uniformly by all the
+#   datapoints: the decision boundary is almost perpendicular to the "Culmen
+#   Length (mm)" feature.
+# - the higher the value of `C` (the weaker the regularization), the more the
+#   decision boundary is influenced by few traing points very close to the
+#   decision boundary, in particular by the misclassified points caused by the
+#   presence of noise in the data, making this classification task non-linearly
+#   separable.
```

Comment:
I don't quite understand what you mean with the last phrase:
> making this classification task non-linearly separable.

Can you please elaborate?

Reply:
Nevermind, I guess you meant noise is non-linearly separable.

I am working locally on my review.

Reply:
Feel free to push commits to the PR.